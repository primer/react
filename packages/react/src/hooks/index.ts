export {useOnOutsideClick} from './useOnOutsideClick'
export type {UseOnOutsideClickSettings, TouchOrMouseEvent} from './useOnOutsideClick'
export {useProvidedRefOrCreate} from './useProvidedRefOrCreate'
export {useOnEscapePress} from './useOnEscapePress'
export {useOpenAndCloseFocus} from './useOpenAndCloseFocus'
export type {UseOpenAndCloseFocusSettings} from './useOpenAndCloseFocus'
export type {AnchoredPositionHookSettings} from './useAnchoredPosition'
export {useAnchoredPosition} from './useAnchoredPosition'
export {useOverlay} from './useOverlay'
export type {UseOverlaySettings} from './useOverlay'
export {useRenderForcingRef} from './useRenderForcingRef'
export {useProvidedStateOrCreate} from './useProvidedStateOrCreate'
export {useMenuInitialFocus} from './useMenuInitialFocus'
export {useMenuKeyboardNavigation} from './useMenuKeyboardNavigation'
export {useMnemonics} from './useMnemonics'
export {useRefObjectAsForwardedRef} from './useRefObjectAsForwardedRef'
export {useId} from './useId'
export {useIsMacOS} from './useIsMacOS'
