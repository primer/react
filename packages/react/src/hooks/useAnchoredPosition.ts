import React from 'react'
import {getAnchoredPosition} from '@primer/behaviors'
import type {AnchorPosition, PositionSettings} from '@primer/behaviors'
import {useProvidedRefOrCreate} from './useProvidedRefOrCreate'
import {useResizeObserver} from './useResizeObserver'
import useLayoutEffect from '../utils/useIsomorphicLayoutEffect'

export interface AnchoredPositionHookSettings extends Partial<PositionSettings> {
  floatingElementRef?: React.RefObject<Element>
  anchorElementRef?: React.RefObject<Element>
  pinPosition?: boolean
}

/**
 * Calculates the top and left values for an absolutely-positioned floating element
 * to be anchored to some anchor element. Returns refs for the floating element
 * and the anchor element, along with the position.
 * @param settings Settings for calculating the anchored position.
 * @param dependencies Dependencies to determine when to re-calculate the position.
 * @returns An object of {top: number, left: number} to absolutely-position the
 * floating element.
 */
export function useAnchoredPosition(
  settings?: AnchoredPositionHookSettings,
  dependencies: React.DependencyList = [],
): {
  floatingElementRef: React.RefObject<Element>
  anchorElementRef: React.RefObject<Element>
  position: AnchorPosition | undefined
} {
  const floatingElementRef = useProvidedRefOrCreate(settings?.floatingElementRef)
  const anchorElementRef = useProvidedRefOrCreate(settings?.anchorElementRef)
  const [position, setPosition] = React.useState<AnchorPosition | undefined>(undefined)
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const [_, setPrevHeight] = React.useState<number | undefined>(undefined)

  const topPositionChanged = (prevPosition: AnchorPosition | undefined, newPosition: AnchorPosition) => {
    return (
      prevPosition &&
      ['outside-top', 'inside-top'].includes(prevPosition.anchorSide) &&
      // either the anchor changed or the element is trying to shrink in height
      (prevPosition.anchorSide !== newPosition.anchorSide || prevPosition.top < newPosition.top)
    )
  }

  const updateElementHeight = () => {
    let heightUpdated = false
    setPrevHeight(prevHeight => {
      // if the element is trying to shrink in height, restore to old height to prevent it from jumping
      if (prevHeight && prevHeight > (floatingElementRef.current?.clientHeight ?? 0)) {
        requestAnimationFrame(() => {
          ;(floatingElementRef.current as HTMLElement).style.height = `${prevHeight}px`
        })
        heightUpdated = true
      }
      return prevHeight
    })
    return heightUpdated
  }

  const updatePosition = React.useCallback(
    () => {
      if (floatingElementRef.current instanceof Element && anchorElementRef.current instanceof Element) {
        const newPosition = getAnchoredPosition(floatingElementRef.current, anchorElementRef.current, settings)
        setPosition(prev => {
          if (settings?.pinPosition && topPositionChanged(prev, newPosition)) {
            const anchorTop = anchorElementRef.current?.getBoundingClientRect().top ?? 0
            const elementStillFitsOnTop = anchorTop > (floatingElementRef.current?.clientHeight ?? 0)

            if (elementStillFitsOnTop && updateElementHeight()) {
              return prev
            }
          }
          return newPosition
        })
      } else {
        setPosition(undefined)
      }
      setPrevHeight(floatingElementRef.current?.clientHeight)
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [floatingElementRef, anchorElementRef, ...dependencies],
  )

  useLayoutEffect(updatePosition, [updatePosition])

  useResizeObserver(updatePosition) // watches for changes in window size
  useResizeObserver(updatePosition, floatingElementRef as React.RefObject<HTMLElement>) // watches for changes in floating element size

  return {
    floatingElementRef,
    anchorElementRef,
    position,
  }
}
