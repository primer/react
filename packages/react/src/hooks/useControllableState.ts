import React from 'react'
import {warning} from '../utils/warning'

type ControllableStateOptions<T> = {
  /**
   * A unique name for the state value.
   */
  name?: string
  /**
   * The default value used for the state. This will be
   * the fallback value used if `value` is not defined.
   * */
  defaultValue: T | (() => T)
  /**
   * A controlled value. Omitting this means that the state is uncontrolled.
   */
  value?: T
  /**
   * An optional function that is called when the value of the state changes.
   * This is useful for communicating to parents of controlled components
   * that the value is requesting to be changed.
   */
  onChange?: (value: T) => void
}

/**
 * This custom hook simplifies the behavior of a component if it has state that
 * can be both controlled and uncontrolled. It functions identical to a
 * useState() hook and provides [state, setState] for you to use. You can use
 * the `onChange` argument to allow updates to the `state` to be communicated to
 * owners of controlled components.
 *
 * Note: This hook will warn if a component is switching from controlled to
 * uncontrolled, or vice-versa.
 */
export function useControllableState<T>({
  name = 'custom',
  defaultValue,
  value,
  onChange,
}: ControllableStateOptions<T>): [T, React.Dispatch<React.SetStateAction<T>>] {
  const [state, internalSetState] = React.useState(value ?? defaultValue)
  const controlled = React.useRef<boolean | null>(null)
  const stableOnChange = React.useRef(onChange)

  React.useEffect(() => {
    stableOnChange.current = onChange
  })

  if (controlled.current === null) {
    controlled.current = value !== undefined
  }

  const setState = React.useCallback(
    (stateOrUpdater: T | ((prevState: T) => T)) => {
      const value =
        typeof stateOrUpdater === 'function'
          ? // @ts-ignore stateOrUpdater is a function
            stateOrUpdater(state)
          : stateOrUpdater

      if (controlled.current === false) {
        internalSetState(value)
      }

      stableOnChange.current?.(value)
    },
    [state],
  )

  React.useEffect(() => {
    const controlledValue = value !== undefined

    // Uncontrolled -> Controlled
    // If the component prop is uncontrolled, the prop value should be undefined
    if (controlled.current === false && controlledValue) {
      warning(
        true,
        'A component is changing an uncontrolled %s component to be controlled. ' +
          'This is likely caused by the value changing to a defined value ' +
          'from undefined. Decide between using a controlled or uncontrolled ' +
          'value for the lifetime of the component. ' +
          'More info: https://reactjs.org/link/controlled-components',
        name,
      )
    }

    // Controlled -> Uncontrolled
    // If the component prop is controlled, the prop value should be defined
    if (controlled.current === true && !controlledValue) {
      warning(
        true,
        'A component is changing a controlled %s component to be uncontrolled. ' +
          'This is likely caused by the value changing to an undefined value ' +
          'from a defined one. Decide between using a controlled or ' +
          'uncontrolled value for the lifetime of the component. ' +
          'More info: https://reactjs.org/link/controlled-components',
        name,
      )
    }
  }, [name, value])

  // eslint-disable-next-line react-hooks/refs
  if (controlled.current === true) {
    return [value as T, setState]
  }

  return [state, setState]
}
