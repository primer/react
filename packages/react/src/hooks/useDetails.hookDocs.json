{
  "name": "useDetails",
  "importPath": "@primer/react",
  "stories": [
    {
      "id": "components-details--default"
    }
  ],
  "parameters": [
    {
      "name": "options",
      "type": "UseDetailsParameters",
      "required": true
    }
  ],
  "returns": {
    "type": "UseDetailsReturn"
  },
  "relatedTypes": [
    {
      "name": "UseDetailsReturn",
      "properties": [
        {
          "name": "open",
          "type": "boolean",
          "description": "Whether the details element is open."
        },
        {
          "name": "setOpen",
          "type": "React.Dispatch<React.SetStateAction<boolean>>",
          "description": "A function to set the open state."
        },
        {
          "name": "getDetailsProps",
          "type": "() => DetailsProps",
          "description": "A function to get the props to attach to the details element."
        }
      ]
    },
    {
      "name": "DetailsProps",
      "properties": [
        {
          "name": "onToggle",
          "type": "(e: React.SyntheticEvent<HTMLElement, Event>) => void",
          "description": "Callback function to handle toggling the details element."
        },
        {
          "name": "open",
          "type": "boolean",
          "description": "Whether the details element is open."
        },
        {
          "name": "ref",
          "type": "React.RefObject<HTMLDetailsElement>",
          "description": "A ref object to attach to the details element."
        }
      ]
    },
    {
      "name": "UseDetailsParameters",
      "properties": [
        {
          "name": "ref",
          "type": "React.RefObject<HTMLDetailsElement>",
          "description": "A ref object to attach to the details element."
        },
        {
          "name": "closeOnOutsideClick",
          "type": "boolean",
          "defaultValue": "false",
          "description": "Whether to close the details element when clicking outside of it."
        },
        {
          "name": "defaultOpen",
          "type": "boolean",
          "defaultValue": "false",
          "description": "Whether the details element is open by default."
        },
        {
          "name": "onClickOutside",
          "type": "(event: MouseEvent) => void",
          "description": "Callback function to handle clicks outside the details element."
        }
      ]
    }
  ]
}
