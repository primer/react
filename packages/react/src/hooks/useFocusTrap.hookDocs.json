{
  "name": "useFocusTrap",
  "importPath": "@primer/react",
  "stories": [
    {"id": "hooks-usefocustrap--custom-initial-focus"},
    {"id": "hooks-usefocustrap--dynamic-focus-trap-contents"},
    {"id": "hooks-usefocustrap--focus-trap"},
    {"id": "hooks-usefocustrap--multiple-focus-traps"},
    {"id": "hooks-usefocustrap--restore-focus"}
  ],
  "parameters": [
    {
      "name": "settings",
      "type": "FocusTrapSettings",
      "required": true,
      "description": "Configuration options for the focus trap."
    },
    {
      "name": "dependencies",
      "type": "Array",
      "required": true,
      "defaultValue": "[]",
      "description": "React.DependencyList"
    }
  ],
  "relatedTypes": [
    {
      "name": "FocusTrapSettings",
      "properties": [
        {
          "name": "containerRef",
          "type": "React.RefObject<HTMLElement>",
          "description": "Ref that will be used for the trapping container. If not provided, one will be created by this hook and returned."
        },
        {
          "name": "initialFocusRef",
          "type": "React.RefObject<HTMLElement>",
          "description": "Ref for the element that should receive focus when the focus trap is first enabled. If not provided, one will be created by this hook and returned. Its use is optional."
        },
        {
          "name": "disabled",
          "type": "boolean",
          "description": "Set to true to disable the focus trap and clean up listeners. Can be re-enabled at any time."
        },
        {
          "name": "restoreFocusOnCleanUp",
          "type": "boolean",
          "description": "If true, when this focus trap is cleaned up, restore focus to the element that had focus immediately before the focus trap was enabled. (Default: false)"
        },
        {
          "name": "returnFocusRef",
          "type": "React.RefObject<HTMLElement>",
          "description": "If passed, when this focus trap is cleaned up, restore focus to this element instead of element with focus immediately before the focus trap was enabled. Overrides `restoreFocusOnCleanUp`."
        }
      ]
    },
    {
      "name": "FocusTrapRefs",
      "properties": [
        {
          "name": "containerRef",
          "type": "React.RefObject<HTMLElement>",
          "description": "Ref that will be used for the trapping container."
        },
        {
          "name": "initialFocusRef",
          "type": "React.RefObject<HTMLElement>",
          "description": "Ref for the child element that will receive focus first."
        }
      ]
    }
  ],
  "returns": {
    "type": "FocusTrapRefs",
    "description": "The refs for the trapping container and the initial focus element."
  }
}
