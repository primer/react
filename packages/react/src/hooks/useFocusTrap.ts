import React from 'react'
import {focusTrap} from '@primer/behaviors'
import {useProvidedRefOrCreate} from './useProvidedRefOrCreate'

export interface FocusTrapHookSettings {
  /**
   * Ref that will be used for the trapping container. If not provided, one will
   * be created by this hook and returned.
   */
  containerRef?: React.RefObject<HTMLElement>

  /**
   * Ref for the element that should receive focus when the focus trap is first enabled. If
   * not provided, one will be created by this hook and returned. Its use is optional.
   */
  initialFocusRef?: React.RefObject<HTMLElement>

  /**
   * Set to true to disable the focus trap and clean up listeners. Can be re-enabled at
   * any time.
   */
  disabled?: boolean

  /**
   * If true, when this focus trap is cleaned up, restore focus to the element that had
   * focus immediately before the focus trap was enabled. (Default: false)
   */
  restoreFocusOnCleanUp?: boolean

  /**
   * If passed, when this focus trap is cleaned up, restore focus to this element instead
   * of element with focus immediately before the focus trap was enabled.
   *
   * Overrides restoreFocusOnCleanUp
   */
  returnFocusRef?: React.RefObject<HTMLElement>
}

/**
 * Hook used to trap focus inside a container. Returns a ref that can be added to the container
 * that should trap focus.
 * @param settings {FocusTrapHookSettings}
 */
export function useFocusTrap(
  settings?: FocusTrapHookSettings,
  dependencies: React.DependencyList = [],
): {containerRef: React.RefObject<HTMLElement>; initialFocusRef: React.RefObject<HTMLElement>} {
  const containerRef = useProvidedRefOrCreate(settings?.containerRef)
  const initialFocusRef = useProvidedRefOrCreate(settings?.initialFocusRef)
  const disabled = settings?.disabled
  const abortController = React.useRef<AbortController>()
  const previousFocusedElement = React.useRef<Element | null>(null)

  // If we are enabling a focus trap and haven't already stored the previously focused element
  // go ahead an do that so we can restore later when the trap is disabled.
  if (!previousFocusedElement.current && !settings?.disabled) {
    previousFocusedElement.current = document.activeElement
  }

  // This function removes the event listeners that enable the focus trap and restores focus
  // to the previously-focused element (if necessary).
  function disableTrap() {
    abortController.current?.abort()
    if (settings?.returnFocusRef && settings.returnFocusRef.current instanceof HTMLElement) {
      settings.returnFocusRef.current.focus()
    } else if (settings?.restoreFocusOnCleanUp && previousFocusedElement.current instanceof HTMLElement) {
      previousFocusedElement.current.focus()
      previousFocusedElement.current = null
    }
  }

  React.useEffect(
    () => {
      if (containerRef.current instanceof HTMLElement) {
        if (!disabled) {
          abortController.current = focusTrap(containerRef.current, initialFocusRef.current ?? undefined)
          return () => {
            disableTrap()
          }
        } else {
          disableTrap()
        }
      }
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [containerRef, initialFocusRef, disabled, ...dependencies],
  )

  return {containerRef, initialFocusRef}
}
