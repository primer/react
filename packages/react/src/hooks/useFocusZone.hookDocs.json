{
  "name": "useFocusZone",
  "importPath": "@primer/react",
  "stories": [
    {"id": "hooks-usefocuszone--active-descendant"},
    {"id": "hooks-usefocuszone--basic-focus-zone"},
    {"id": "hooks-usefocuszone--changing-subtree"},
    {"id": "hooks-usefocuszone--custom-focus-movement"},
    {"id": "hooks-usefocuszone--focus-in-strategy"},
    {"id": "hooks-usefocuszone--focus-out-behavior"},
    {"id": "hooks-usefocuszone--nested-zones"},
    {"id": "hooks-usefocuszone--special-situations"}
  ],
  "parameters": [
    {
      "name": "settings",
      "type": "FocusZoneHookSettings",
      "description": "Configuration options for the focus zone."
    },
    {
      "name": "dependencies",
      "type": "React.DependencyList",
      "required": true,
      "defaultValue": "[]"
    }
  ],
  "relatedTypes": [
    {
      "name": "FocusZoneHookSettings",
      "properties": [
        {
          "name": "containerRef",
          "type": "React.RefObject<HTMLElement>",
          "description": "Optional ref for the container that holds all elements participating in arrow key focus. If one is not passed, we will create one for you and return it from the hook."
        },
        {
          "name": "activeDescendantFocus",
          "type": "boolean | React.RefObject<HTMLElement>",
          "description": "If using the 'active descendant' focus pattern, pass `true` or a ref to the controlling element. If a ref object is not passed, we will create one for you."
        },
        {
          "name": "disabled",
          "type": "boolean",
          "description": "Set to true to disable the focus zone and clean up listeners. Can be re-enabled at any time."
        },
        {
          "name": "focusOutBehavior",
          "type": "'stop' | 'wrap'",
          "description": "Defines the behavior when focus moves out of the focus zone. Can be 'stop' or 'wrap'."
        },
        {
          "name": "getNextFocusable",
          "type": "(direction: 'previous' | 'next' | 'start' | 'end', from: Element | undefined, event: KeyboardEvent) => HTMLElement | undefined",
          "description": "Function to determine the next focusable element in the given direction."
        },
        {
          "name": "focusableElementFilter",
          "type": "(element: HTMLElement) => boolean",
          "description": "Function to filter which elements are focusable."
        },
        {
          "name": "bindKeys",
          "type": "enum FocusKeys {\nArrowHorizontal = 1,\nArrowVertical = 2,\nJK = 4,\nHL = 8,\nHomeAndEnd = 16,\nPageUpDown = 256,\nWS = 32,\nAD = 64,\nTab = 128,\nBackspace = 512,\nArrowAll = 3,\nHJKL = 12,\nWASD = 96,\nAll = 511\n}",
          "defaultValue": "FocusKeys.ArrowVertical",
          "description": "Keys that should be used to move focus within the focus zone."
        },
        {
          "name": "abortSignal",
          "type": "AbortSignal",
          "description": "Signal to abort the focus zone."
        },
        {
          "name": "activeDescendantControl",
          "type": "HTMLElement",
          "description": "Element that controls the active descendant."
        },
        {
          "name": "focusInStrategy",
          "type": "'first' | 'closest' | 'previous' | ((previousFocusedElement: Element) => HTMLElement | undefined)",
          "description": "Strategy to determine which element receives focus when entering the focus zone."
        },
        {
          "name": "preventScroll",
          "type": "boolean",
          "description": "Prevents scrolling when focusing elements."
        }
      ]
    },
    {
      "name": "FocusZoneRefs",
      "properties": [
        {
          "name": "containerRef",
          "type": "React.RefObject<HTMLElement>",
          "description": "Ref that will be used for the element with the children focus will be moved between."
        },
        {
          "name": "initialFocusRef",
          "type": "React.RefObject<HTMLElement>",
          "description": "Ref for the child element that will receive focus first."
        }
      ]
    },
    {
      "name": "FocusKeys (enum)",
      "properties": [
        {
          "name": "ArrowHorizontal",
          "type": "1",
          "description": "Left and right arrow keys. Prevents default behavior of scrolling where applicable."
        },
        {
          "name": "ArrowVertical",
          "type": "2",
          "description": "Up and down arrow keys. Prevents default behavior of scrolling where applicable."
        },
        {
          "name": "ArrowAll",
          "type": "3",
          "description": "Left, right, up, and down arrow keys. Prevents default behavior of scrolling where applicable."
        },
        {
          "name": "HomeAndEnd",
          "type": "16",
          "description": "Home and end keys. Causes focus to jump to the first or last focusable item in the container. Does not move focus if the currently-focused element is a text box."
        },
        {
          "name": "PageUpDown",
          "type": "256",
          "description": "Works the same as the Home and End keys. Advisable only when supplying a custom callback that supports paging."
        },
        {
          "name": "JK",
          "type": "4",
          "description": "J moves focus to the next item, K moves to the previous. Does not move focus if the currently-focused element is a text box. [Originally from](https://catonmat.net/why-vim-uses-hjkl-as-arrow-keys) the vi keybindings"
        },
        {
          "name": "HL",
          "type": "8",
          "description": "H moves focus to the previous item, L moves to the next. Does not move focus if the currently-focused element is a text box. [Originally from](https://catonmat.net/why-vim-uses-hjkl-as-arrow-keys) the vi keybindings"
        },
        {
          "name": "WS",
          "type": "32",
          "description": "W moves focus to the previous item, S moves to the next. Does not move focus if the currently-focused element is a text box."
        },
        {
          "name": "AD",
          "type": "64",
          "description": "A moves focus to the previous item, D moves to the next. Does not move focus if the currently-focused element is a text box."
        },
        {
          "name": "HJKL",
          "type": "12"
        },
        {
          "name": "WASD",
          "type": "96"
        },
        {
          "name": "Tab",
          "type": "128",
          "description": "Unlike other keys, the Tab key will always allow movement outside of the focus zone (use the Focus Trap behavior to prevent this). Tab moves to the next item, Shift+Tab moves to the previous."
        },
        {
          "name": "Backspace",
          "type": "512",
          "description": "Moves focus to the previous item. Does not move focus if the currently-focused element is a text box."
        },
        {
          "name": "All",
          "type": "511",
          "description": "All keys listed above are bound."
        }
      ]
    }
  ],
  "returns": {
    "type": "FocusZoneRefs",
    "description": "The refs for the trapping container and the initial focus element."
  }
}
