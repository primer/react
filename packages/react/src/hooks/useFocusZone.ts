import React, {useEffect} from 'react'
import {focusZone} from '@primer/behaviors'
import type {FocusZoneSettings} from '@primer/behaviors'
import {useProvidedRefOrCreate} from './useProvidedRefOrCreate'
export {FocusKeys} from '@primer/behaviors'
export type {Direction} from '@primer/behaviors'

export interface FocusZoneHookSettings extends Omit<FocusZoneSettings, 'activeDescendantControl'> {
  /**
   * Optional ref for the container that holds all elements participating in arrow key focus.
   * If one is not passed, we will create one for you and return it from the hook.
   */
  containerRef?: React.RefObject<HTMLElement>

  /**
   * If using the "active descendant" focus pattern, pass `true` or a ref to the controlling
   * element. If a ref object is not passed, we will create one for you.
   */
  activeDescendantFocus?: boolean | React.RefObject<HTMLElement>

  /**
   * Set to true to disable the focus zone and clean up listeners. Can be re-enabled at
   * any time.
   */
  disabled?: boolean
}

export function useFocusZone(
  settings: FocusZoneHookSettings = {},
  dependencies: React.DependencyList = [],
): {containerRef: React.RefObject<HTMLElement>; activeDescendantControlRef: React.RefObject<HTMLElement>} {
  const containerRef = useProvidedRefOrCreate(settings.containerRef)
  const useActiveDescendant = !!settings.activeDescendantFocus
  const passedActiveDescendantRef =
    typeof settings.activeDescendantFocus === 'boolean' || !settings.activeDescendantFocus
      ? undefined
      : settings.activeDescendantFocus
  const activeDescendantControlRef = useProvidedRefOrCreate(passedActiveDescendantRef)
  const disabled = settings.disabled
  const abortController = React.useRef<AbortController>()

  useEffect(
    () => {
      if (
        containerRef.current instanceof HTMLElement &&
        (!useActiveDescendant || activeDescendantControlRef.current instanceof HTMLElement)
      ) {
        if (!disabled) {
          const vanillaSettings: FocusZoneSettings = {
            ...settings,
            activeDescendantControl: activeDescendantControlRef.current ?? undefined,
          }
          abortController.current = focusZone(containerRef.current, vanillaSettings)
          return () => {
            abortController.current?.abort()
          }
        } else {
          abortController.current?.abort()
        }
      }
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [disabled, ...dependencies],
  )

  return {containerRef, activeDescendantControlRef}
}
