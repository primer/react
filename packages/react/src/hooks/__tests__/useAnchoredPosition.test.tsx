import {render, waitFor} from '@testing-library/react'
import {it, expect, vi} from 'vitest'
import React from 'react'
import {useAnchoredPosition} from '../../hooks/useAnchoredPosition'

const Component = ({callback}: {callback: (hookReturnValue: ReturnType<typeof useAnchoredPosition>) => void}) => {
  const floatingElementRef = React.useRef<HTMLDivElement>(null)
  const anchorElementRef = React.useRef<HTMLDivElement>(null)
  callback(useAnchoredPosition({floatingElementRef, anchorElementRef}))
  return (
    <div style={{position: 'absolute'}}>
      <div
        style={{position: 'absolute', top: '20px', left: '20px', height: '50px', width: '50px'}}
        ref={floatingElementRef}
      />
      <div ref={anchorElementRef} />
    </div>
  )
}

it('should should return a position', async () => {
  const cb = vi.fn()
  render(<Component callback={cb} />)

  await waitFor(() => {
    expect(cb).toHaveBeenCalledTimes(2)
    expect(cb.mock.calls[1][0]['position']).toMatchInlineSnapshot(`
    {
      "anchorAlign": "start",
      "anchorSide": "outside-bottom",
      "left": 0,
      "top": 4,
    }
  `)
  })
})
