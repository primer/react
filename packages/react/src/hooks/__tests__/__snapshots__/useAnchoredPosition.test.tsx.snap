// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should should return a position 1`] = `
{
  "anchorAlign": "center",
  "anchorSide": "outside-bottom",
  "left": 8.550800000000002,
  "top": 24.5,
}
`;

exports[`should should return a position 2`] = `
{
  "anchorAlign": "center",
  "anchorSide": "outside-bottom",
  "left": 8.554699999999997,
  "top": 28,
}
`;
